APMrover2/sailboat.cpp

```diff
@@ -67,3 +69,131 @@ float Rover::sailboat_get_VMG() const
+{
+    return _sailboat_tacking;
+}
+
+// returns true if sailboat should take a indirect navigation route to go upwind
+// desired_heading should be in centi-degrees
+bool Rover::sailboat_use_indirect_route(float desired_heading_cd) const
+{
+    if (!g2.motors.has_sail()) {
+        return false;
+    }
+
+    // convert desired heading to radians
+    const float desired_heading_rad = radians(desired_heading_cd * 0.01f);
+
+    // check if desired heading is in the no go zone, if it is we can't go direct
+    return fabsf(wrap_PI(g2.windvane.get_absolute_wind_direction_rad() - desired_heading_rad)) <= radians(g2.sail_no_go);
+}
+
+// if we can't sail on the desired heading then we should pick the best heading that we can sail on
+// this function assumes the caller has already checked sailboat_use_indirect_route(desired_heading_cd) returned true
+float Rover::sailboat_calc_heading(float desired_heading_cd)
+{
+    if (!g2.motors.has_sail()) {
+        return desired_heading_cd;
+    }
+
+    bool should_tack = false;
+
+    // check for user requested tack
+    uint32_t now = AP_HAL::millis();
+    if (_sailboat_auto_tack_request_ms != 0) {
+        // set should_tack flag is user requested tack within last 0.5 sec
+        should_tack = ((now - _sailboat_auto_tack_request_ms) < 500);
+        _sailboat_auto_tack_request_ms = 0;
+    }
+
+    // calculate left and right no go headings looking upwind
+    const float left_no_go_heading_rad = wrap_2PI(g2.windvane.get_absolute_wind_direction_rad() + radians(g2.sail_no_go));
+    const float right_no_go_heading_rad = wrap_2PI(g2.windvane.get_absolute_wind_direction_rad() - radians(g2.sail_no_go));
+
+    // calculate current tack, Port if heading is left of no-go, STBD if right of no-go
+    Sailboat_Tack current_tack;
+    if (is_negative(g2.windvane.get_apparent_wind_direction_rad())) {
+        current_tack = Tack_Port;
+    } else {
+        current_tack = Tack_STBD;
+    }
+
+    // trigger tack if cross track error larger than waypoint_overshoot parameter
+    // this effectively defines a 'corridor' of width 2*waypoint_overshoot that the boat will stay within
+    if ((fabsf(rover.nav_controller->crosstrack_error()) >= g.waypoint_overshoot) && !is_zero(g.waypoint_overshoot) && !sailboat_tacking()) {
+        // make sure the new tack will reduce the cross track error
+        // if were on starboard tack we are traveling towards the left hand boundary
+        if (is_positive(rover.nav_controller->crosstrack_error()) && (current_tack == Tack_STBD)) {
+            should_tack = true;
+        }
+        // if were on port tack we are traveling towards the right hand boundary
+        if (is_negative(rover.nav_controller->crosstrack_error()) && (current_tack == Tack_Port)) {
+            should_tack = true;
+        }
+    }
+
+    // if tack triggered, calculate target heading
+    if (should_tack) {
+        gcs().send_text(MAV_SEVERITY_INFO, "Sailboat: Tacking");
```

Comment:
ideally we should remove this with a real mavlink message. Maybe we should define one for sailboat

Reply:
txs for looking at this.  I'm happy to remove this message if people prefer.  I generally think we should try to avoid sending text messages when things are working as they should.

Reply:
I really just added this to be able to tell if it was trying to tack but failing in very light wind, more of a debug than anything else. I figured it was the equivalent of the quad-plane transition messages.

Reply:
let's keep it for now if that kind of message doens't appear much